feat: Native Dockerfile

aht007 opened the pull request:
ISSUE: https://github.com/openedx/edx-analytics-data-api/issues/579
This PR is part of effort aimed at removing Ansible based configurations and replacing them with `Dockerfile`. Currently Devstack Docker images are built using `Ansible based configurations` in the `configurations` repository. Through this effort we will make sure that the Repo has its own `Dockerfile` which has all the necessary configurations to setup small production and dev environments.

**Steps to run this Image with Devstack:**

- Build the Image locally first using the target `dev` i.e. `docker build -t analytics_api_dev --target dev .`
- Checkout to the branch in this PR of `devstack` https://github.com/openedx/devstack/pull/986
- Run `make dev.provision.analytics_api` to run the provisioning script for the new image.
- Run `make dev.up.analytics_api` in the terminal while in the devstack directory.

aht007 committed:
feat: analyticsapi config for native image

aht007 committed:
fix: fix

aht007 commented:
@openedx/masters-devs-cosmonauts I can see that the current Dockerfile of this repository is being used for CI purposes and is also being pushed to `openedx` docker hub account. I would need your team's guidance as to what extent that image is useful and do we need to maintain that image as well or we can just retire that and use our new docker image for CI testing as well.
Besides that I would also like for your team to take a look at the changes in this PR.

schenedx commented on requirements/dev.in:
```diff
@@ -1,3 +1,4 @@
 # Local development dependencies go here
 
 -r base.in                          # Core dependencies of edx-analytics-data-api
+mysqlclient
```
Can you explain why this is needed now with a native docker file? Is it the case that Ansible was taking care of this dependency?

aht007 replied:
Yeah, in ansible the production file was also being installed along with the dev requirements file. That's not what we want now as dev target should only install dev requirements file, but when I tried to do this the container threw errors while connecting to mysql because `mysqlclient` was missing. That's why I added it to `dev.in` instead of installing the `production.txt` in dev image.

schenedx replied:
Thanks for that explanation. This make sense.

schenedx commented on Dockerfile:
```diff
@@ -1,34 +1,77 @@
-FROM app as newrelic
-RUN pip install newrelic
-CMD newrelic-admin run-program gunicorn --bind=0.0.0.0:8100 --workers 2 --max-requests=1000 -c /edx/app/analytics_api/analytics_data_api/docker_gunicorn_configuration.py analyticsdataserver.wsgi:application
+CMD while true; do python ./manage.py runserver 0.0.0.0:8110; sleep 2; done
```
We are no longer hooking up with newrelic-admin with this change. Why would you want to leave that out?

aht007 replied:
We don't use new relic in `devstack`. As for the current case I haven't added it to prod target as well since according to our discussion with Jeremy we don't want to make `new relic` a necessary part of our base images by default. We can make a separate target for that in our Dockerfile or leave it to our orchestration system(k8s or docker-compose) to do those things for us.

schenedx replied:
Got it. Do you have wiki page documents talk about this? I'd like this point to be written down as a policy for layering newrelic to our apps.

schenedx commented on .github/workflows/push-docker-images.yml:
```diff
@@ -0,0 +1,49 @@
+          repository: edxops/analytics-api-dev
+          tags: ${{ steps.get-tag-name.outputs.result }},${{ github.sha }}
+
+      # - name: Build and push prod Docker image
```
Why is this commented out? Is it the case we don't really push this image to any repositories?

aht007 replied:
I have commented that out on purpose. Since our initial target is `devstack` so I am just pushing the dev images for now. Also the old Ansible based `devstack` image is pushed to `edxops/analytics_api` repository and enabling this push will override that image too. For smooth transition, I am pushing the Native Images to new repository(`analytics_api_dev`) and keeping the old image intact so if we want to switch over to older image due to some issues we always can.

schenedx replied:
If you are doing this on purpose, then the whole section should be wrapped with comments to help others understand why this is left in the code, but commented out. Please provide the explanation in the code to help others understand the desire for the smooth transition.

schenedx reviewed (commented):
I got inline comments.

alangsto commented on Dockerfile:
```diff
@@ -1,34 +1,77 @@
 RUN apt update && \
-  apt-get install -y software-properties-common && \
-  apt-add-repository -y ppa:deadsnakes/ppa && apt-get update && \
-  apt install -y git-core language-pack-en python3.8-dev python3.8-venv libmysqlclient-dev libffi-dev libssl-dev build-essential gettext openjdk-8-jdk && \
```
it looks like not all of these packages were copied over in the new format, is that intentional (i.e. `libffi-dev`)

aht007 replied:
Yeah thats intentional. I couldn't find any use of these packages e.g `gettext`, `libffi-dev`, `git-core`, `openjdk` and others so I removed them.

schenedx replied:
@aht007 Could you explain to us how you arrived at the conclusion that these are not used? Did you bring up the service and test through the newly built docker image?

aht007 replied:
It is both testing the new image as well as learning from other images as well. These packages were also present in Insights' Ansible roles but I removed them from there as well since I couldn't find their usage. Also as per our knowledge many of these packages are present in Ansible roles and docker files for historical reasons but they aren't used now so we are better off without them.
I have tried bringing up the container and testing the image and it looked fine but if someone from your team knows about any of the removed package's usage then I will be happy to put it back there.

schenedx reviewed (approved):
👍

aht007 committed:
fix: remove redundant image and test new pull

aht007 committed:
fix: remove test push trigger

aht007 committed:
test: run tests on new image

aht007 commented:
@schenedx Regarding the old docker publish workflow, I searched for that image's usage and I couldn't find it anywhere, not in this repo and not in the `openedx` org at all. That workflow was added as a part of CI overhaul when we shifted to Github Actions and I think that image was being pushed for just historical reasons without any usage. I have removed that workflow and related commands. Also the CI was using old Ansible based image which I have replaced with the newer image and CI is running fine with that(I tested the push in this PR so the new image is available on edxops). 
I would again need one final review from your side on this PR as well this [PR of devstack](https://github.com/openedx/devstack/pull/986) which introduces the newer image usage.

schenedx commented:
@aht007 Reviewed your devstack PR as well as re-review on this one. Still LGTM.
Feel free to merge.
